proposals/0000-integrations.md

```diff
@@ -0,0 +1,171 @@
+- Nuxt: [1 step](https://partytown.builder.io/nuxt)
+
+Tailwind suffers from a similar difficult setup story in Astro.
+
+# Detailed design
+
+**Background:** This API was reached through weeks of experimentation of different designs (see alternatives below). To test the work, I designed and build the following integrations, which are useful for illustrating this RFC: 
+
+- **Renderers:** [`lit`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/lit/index.js), [`svelte`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/svelte/index.js), [`react`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/react/index.js), [`preact`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/preact/index.js), [`vue`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/vue/index.js), [`solid`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/solid/index.js)
+- **Libraries:** [`tailwind`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/tailwind/index.js), [`partytown`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/partytown/index.js), [`turbolinks`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/turbolinks/index.js)
+- **Features:** [`sitemap`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/sitemap/index.js)
+
+## Integration Usage API
+
+```js
+// astro.config.js
+export default ({
+	integrations: [
+		import('@astrojs/vue'),
+        // or, with options
+		[import('@astrojs/vue'), {/* options */}],
+        // or, as a static ESM import at the top of the file
+		vuePlugin({/* options */})
+	],
+});
+```
+
+## Integration Author API
+
+```ts
+export interface AstroIntegration {
+	name: string;
+	hooks: {
+		/** SETUP */
+		/** Called on astro startup, lets you modify config */
+		'astro:config:setup': (options: ConfigSetupOptions) => void | Promise<void>;
+		/** Called after config is finalized, lets you store config object for later */
+		'astro:config:done': (options: { config: Readonly<AstroConfig> }) => void | Promise<void>;
+
+		/** DEV */
+		/** Called on server setup, lets you modify the server */
+		'astro:server:setup': (options: { server: vite.ViteDevServer }) => void | Promise<void>;
+		/** Called on server startup, lets you read the dev server address/URL */
+		'astro:server:start': (options: { address: AddressInfo }) => void | Promise<void>;
+		/** Called on server exit */
+		'astro:server:done': () => void | Promise<void>;
+
+		/** BUILD */
+		/** Called on build start, lets you modify the server */
+		'astro:build:start': () => void | Promise<void>;
+		/** Called on build done, lets you read metadata about the build */
+		'astro:build:done': (options: { pages: string[]; dir: URL }) => void | Promise<void>;
```

Comment:
I think this hook could probably be used by an `Image` component to write optimized images to the dist folder, does that sound right?

Reply:
Not sure how this would work with the dev server and SSR though? Assuming this array of `pages` is the _raw_, compiled HTML, I don't think we could allow DOM crawling and parsing on SSR in particular.

I could see us dividing 2 separate hooks to address this: `astro:compile` and `astro:render`.
- Compile: Gives you plain metadata about each page like pathname, dir, and potentially a metadata object if the meta proposal #135 is considered. Image optimization would _ideally_ happen here since SSR routes would also get picked up at build time.
- Render: Gives you the rendered HTML output. Image optimization _can_ happen here, as long as you're using optimization libraries fast enough for serverless requests.

Reply:
@bholmesdev These hooks only run doing `astro build` and `astro dev`, not in SSR. The sort of image optimizations I'm thinking of could probably only be done for an SSG site.

Reply:
It's hard to say without us having a clear implementation, but as a concept I really like the idea of building our own features as integrations as a way to dogfood and keep improving the integration API for others.